rfcs/text/0005_route_handler.md

```diff
@@ -0,0 +1,187 @@
+Rather `Route Handler` in New platform has the next signature:
+```typescript
+type RequestHandler = (
+  context: Context,
+  request: KibanaRequest,
+  t: KibanaResponseToolkit
+) => KibanaResponse | Promise<KibanaResponse>;
+```
+and accepts next Kibana specific parameters as arguments:
+- context: [Context](https://github.com/elastic/kibana/blob/master/rfcs/text/0003_handler_interface.md). A handler context contains core service and plugin functionality already scoped to the incoming request.
+- request: [KibanaRequest](https://github.com/elastic/kibana/blob/master/src/core/server/http/router/request.ts). An immutable representation of the incoming request details, such as body, parameters, query, url and route information. Note: you **must** to specify route schema during route declaration to have access to `body, parameters, query` in the request object. You  cannot extend KibanaRequest with arbitrary data nor remove any properties from it.
+```typescript
+interface KibanaRequest {
+  url: url.Url;
+  headers: Record<string, string | string [] | undefined>;
+  params?: Record<string, any>;
+  body?: Record<string, any>;
+  query?: Record<string, any>;
+  route: {
+    path: string;
+    method: 'get' | 'post' | ...
+    options: {
+      authRequired: boolean;
+      tags: string [];
+    }
+  }
+}
+```
+-  t: [KibanaResponseToolkit](https://github.com/elastic/kibana/blob/master/src/core/server/http/router/response.ts#L27)
+Provides a set of pre-configured methods to respond to an incoming request. It is expected that handler **always** returns a result of one of `KibanaResponseToolkit` methods as an output:
+```typescript
+interface KibanaResponseToolkit {
+  [method:string]: (...params: any) => KibanaResponse
+}
+router.get(...,
+  (context: Context, request: KibanaRequest, t: KibanaResponseToolkit): KibanaResponse => {
+   return t.ok();
+   // or
+   return t.redirected('/url');
+   // or
+   return t.badRequest(error);
+);
+```
+*KibanaResponseToolkit*  methods allow an end user to adjust the next response parameters:
+- Body. Supported values:`undefined | string | JSONValue | Buffer | Stream`.
+- Status code. 
+- Headers. Supports adjusting [known values](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v10/http.d.ts#L8) and attaching [custom values as well](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v10/http.d.ts#L67)
+
+Other response parameters, such as `etag`, `MIME-type`, `bytes` that used in the Legacy platform could be adjusted via Headers.
+
+The router handler doesn't expect that logic inside can throw or return something different from `KibanaResponse`. In this case, Http service will respond with `Server error` to prevent exposure of internal logic details.
+
+#### KibanaResponseToolkit methods
+Basic primitives:
+```typescript
+type HttpResponsePayload = undefined | string | JSONValue | Buffer | Stream;
```

Comment:
I assume a more detailed design of the response payloads for various types of responses is outside the scope of this RFC?

It would be nice to provide some consistency in the payload, for example requiring a `message` in the payload for a `400` error. The response toolkit DSL would enable us to do this; I'm mostly wondering where/when this discussion would take place. (Or more broadly, how much of this we'd actually plan to enforce in core).

Reply:
I'd start with more generic definitions outlined in this PR and then tighten them case by case.